### DISCUSSION TITLE
How to Flash firmware that includes Conditional Rules

### DISCUSSION THREAD

**Discussion by jackpotcityco** [arendst/Tasmota#19747] [Q&A] [answered]
Hello,

**My Current Firmware**
```
Program Version	13.1.0.4(tasmota)
Build Date & Time	2023-10-09T14:35:26
Core/SDK Version	2_7_4_9/2.2.2-dev(38a443e)
```
**My ESP8266 as it shows in Tasmota configuration now**
```
ESP Chip Id	1574764 (ESP8266EX)
Flash Chip Id	0x16405E (DOUT)
Flash Size	4096 KB
Program Flash Size	1024 KB
Program Size	639 KB
Free Program Space	364 KB
Free Memory	24.1 KB
```
**I have Visual Studio Code with Extension PlatformIO installed**
...........................................................................................................................................................................................................................

I was reading in the documentation about:
![image](https://github.com/arendst/Tasmota/assets/47510843/2155558b-269e-4a6a-b98b-0d509109e7ce)

What I would try to do is to install a new Firmware which would support: `Support IF, ELSEIF, ELSE`

I have tried to read everywhere but are not exactly sure how to flash this correctly. Below are the things I have done so far:

**FIRST APPROACH:**
1. I have cloned this into the folder: `C:\Tasmota`
`git clone https://github.com/arendst/Tasmota.git`

2. I have create a copy of `C:\Tasmota\tasmota\user_config_override_sample.h` file with the name: `C:\Tasmota\tasmota\user_config_override.h`
3. Inside `C:\Tasmota\tasmota\user_config_override.h` I have added the line: ` #define SUPPORT_IF_STATEMENT`

**SECOND APPROACH:**
1. If I understand correct, my ESP8266EX board has 4096 KB which would enable this approach which could perheps be the one that is simplier?
On the link: `https://ota.tasmota.com/tasmota/` I have downloaded the `http://ota.tasmota.com/tasmota/tasmota-4M.bin.gz`
as the image above suggest already support the `Conditional Rules`. 
2. When downloading this, I only got a folder like this: `tasmota-4M.bin/tasmota-4M.bin` with that only file in it.

Now, I am not sure how to continue from here exactly. Should I use `Visual Studio Code` to flash this firmware, - or I did also see that there could be a web page where it was possible to flash firmware but not sure if this one can be flashed there and would that only be this file?

Preferable, perheps flashing the `tasmota-4M.bin` would be a good idéa? But I don't know how to proceed from here. Is that the only file which is needed to do it? I would be open for any tool how to do it.

Thank you!

**Comment by barbudor** [comment, marked as answer]
3rd approach, the lazy one
Connect your board to usb and open https://tasmota.github.io/install/ in your browser
Select "Tasmota for 4MB flash" and proceed following the instructions


1. On ESP8266 GZ files are not meant to be uncompressed on your PC but directly used for upgrading through OTA. Tasmota un compress the file on the fly. BIN files are exclusively for serial flashing. 
2. If your board already had been flashed with a tasmota binary for 1MB flash, you cannot OTA to a 4MB image (or same the other way round). You need to serial flash
3. Once you have a 4MB binary running you can upgrade OTA by either
a) enter the Url http://ota.tasmota.com/tasmota/tasmota-4M.bin.gz in the "OTA url" section of the firmware upgrade, and let the magic happen
b) download tasmota-minimal.bin.gz on your computer and upgrade to that variant using "upgrade by file upload". Once rebooted in Tasmota-minimal, follow the same procedure to upgrade to another GZ 4MB image whether you have compiled it yourself or downloaded. 

Always use GZ binary as-is for upgrades 
BIN are for serial flashing only

Binaries larger than 300kb have to be flashed in 2 steps, first going through tasmota minimal. The Ora url method does that for you automatically under the hood

**Reply by jackpotcityco** [reply]
That did work out well!

1. I used the https://tasmota.github.io/install/ and selected: **"Tasmota for 4MB flash"**
2. I choosed to erase the ESP8266EX before proceeding as I had a firmware on already.
3. I then entered the Url http://ota.tasmota.com/tasmota/tasmota-4M.bin.gz in the "OTA url" section of the firmware upgrade and it upgraded successfully here too.

Now it looks like this in the Web Interface:
```
Program Version	13.1.0.4(tasmota-4M)
Build Date & Time	2023-10-15T12:52:50
```

Thank you for your help!
